<a href="https://colab.research.google.com/github/neurosxx/Descriptive-Stats-Phase-2-/blob/main/Filtering_Data_with_SQL_Lab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Filtering Data with SQL - Lab





## Introduction

NASA wants to go to Mars! Before they build their rocket, NASA needs to track information about all of the planets in the Solar System. In this lab, you'll practice querying the database with various `SELECT` statements. This will include selecting different columns and implementing other SQL clauses like `WHERE` to return the data desired.

<img src="https://raw.githubusercontent.com/learn-co-curriculum/dsc-filtering-lab-v2-4/master/images/planets.png" alt="image of solar system" width="600">

## Objectives

You will practice the following:

* Retrieve a subset of records from a table using a `WHERE` clause
* Filter results using conditional operators such as `BETWEEN`, `IS NULL`, and `LIKE`
* Apply an aggregate function to the result of a filtered query

## Connecting to the Database

To get started, import `sqlite3` as well as `pandas` for conveniently displaying results. Then, connect to the SQLite database located at `planets.db`.

In [1]:
import pandas as pd
import sqlite3
conn = sqlite3.connect('planets.db')
pd.read_sql("""
SELECT *
FROM planets;
""", conn)

id     name       color  num_of_moons    mass  rings
0   1  Mercury        gray             0    0.55      0
1   2    Venus      yellow             0    0.82      0
2   3    Earth        blue             1    1.00      0
3   4     Mars         red             2    0.11      0
4   5  Jupiter      orange            68  317.90      0
5   6   Saturn       hazel            62   95.19      1
6   7   Uranus  light blue            27   14.54      1
7   8  Neptune   dark blue            14   17.15      1

## Database Schema

This database contains a single table, `planets`. This is the schema:

```
CREATE TABLE planets (
  id INTEGER PRIMARY KEY,
  name TEXT,
  color TEXT,
  num_of_moons INTEGER,
  mass REAL,
  rings BOOLEAN
);
```

The data looks something like this:

| id | name    | color      | num_of_moons | mass   | rings |
| -- | ------- | ---------- | ------------ | ------ | ----- |
| 1  | Mercury | gray       | 0            | 0.55   | FALSE |
| 2  | Venus   | yellow     | 0            | 0.82   | FALSE |
| 3  | Earth   | blue       | 1            | 1.00   | FALSE |
| 4  | Mars    | red        | 2            | 0.11   | FALSE |
| 5  | Jupiter | orange     | 67           | 317.90 | FALSE |
| 6  | Saturn  | hazel      | 62           | 95.19  | TRUE  |
| 7  | Uranus  | light blue | 27           | 14.54  | TRUE  |
| 8  | Neptune | dark blue  | 14           | 17.15  | TRUE  |

## SQL Queries

Write SQL queries for each of the statements below using the same pandas wrapping syntax from the previous lesson.

### 1. Select just the `name` and `color` of each planet

In [2]:
pd.read_sql("""
SELECT name, color
FROM planets;
""", conn)

name       color
0  Mercury        gray
1    Venus      yellow
2    Earth        blue
3     Mars         red
4  Jupiter      orange
5   Saturn       hazel
6   Uranus  light blue
7  Neptune   dark blue

### 2. Select all columns for each planet whose `num_of_moons` is 0

In [3]:
pd.read_sql("""
SELECT *
FROM planets
WHERE num_of_moons = 0
""", conn)

id     name   color  num_of_moons  mass  rings
0   1  Mercury    gray             0  0.55      0
1   2    Venus  yellow             0  0.82      0

\### 3. Select the `name` and `mass` of each planet whose `name` has exactly 7 letters

In [4]:
pd.read_sql("""
SELECT name AS Seven_letter_planets, mass
FROM planets
WHERE LENGTH(name) = 7;
""", conn)

Seven_letter_planets    mass
0              Mercury    0.55
1              Jupiter  317.90
2              Neptune   17.15

### 4. Select all columns for each planet whose `mass` is greater than 1.00

In [5]:
pd.read_sql("""
SELECT name AS Seven_letter_planets, mass
FROM planets
WHERE mass > 1.00;
""", conn)

Seven_letter_planets    mass
0              Jupiter  317.90
1               Saturn   95.19
2               Uranus   14.54
3              Neptune   17.15

### 5. Select the `name` and `mass` of each planet whose `mass` is less than or equal to 1.00

In [6]:
pd.read_sql("""
SELECT name AS Seven_letter_planets, mass
FROM planets
WHERE mass <= 1.00;
""", conn)

Seven_letter_planets  mass
0              Mercury  0.55
1                Venus  0.82
2                Earth  1.00
3                 Mars  0.11

### 6. Select the `name` and `mass` of each planet whose `mass` is between 0 and 50

In [7]:
pd.read_sql("""
SELECT name, mass
FROM planets
WHERE mass BETWEEN 0 AND 50;
""", conn)

name   mass
0  Mercury   0.55
1    Venus   0.82
2    Earth   1.00
3     Mars   0.11
4   Uranus  14.54
5  Neptune  17.15

### 7. Select all columns for planets that have at least one moon and a `mass` less than 1.00

***Hint:*** You can use `AND` to chain together two conditions in SQL, similar to `and` in Python

In [8]:
pd.read_sql("""
SELECT name, mass, num_of_moons
FROM planets
WHERE (num_of_moons) >= 1 AND mass < 1.00;
""", conn)

name  mass  num_of_moons
0  Mars  0.11             2

### 8. Select the `name` and `color` of planets that have a `color` containing the string "blue"

In [9]:
pd.read_sql("""
SELECT name, color
FROM planets
WHERE color = 'blue';
""", conn)

name color
0  Earth  blue

In [10]:
pd.read_sql("""
SELECT *
FROM planets;
""", conn)

id     name       color  num_of_moons    mass  rings
0   1  Mercury        gray             0    0.55      0
1   2    Venus      yellow             0    0.82      0
2   3    Earth        blue             1    1.00      0
3   4     Mars         red             2    0.11      0
4   5  Jupiter      orange            68  317.90      0
5   6   Saturn       hazel            62   95.19      1
6   7   Uranus  light blue            27   14.54      1
7   8  Neptune   dark blue            14   17.15      1

### 9. Select the count of planets that don't have rings as `planets_without_rings`

Note: even though the schema states that `rings` is a `BOOLEAN` and the example table shows values `TRUE` and `FALSE`, SQLite does not actually support booleans natively. From the [documentation](https://www.sqlite.org/datatype3.html#boolean_datatype):

>



> SQLite does not have a separate Boolean storage class. Instead, Boolean values are stored as integers 0 (false) and 1 (true).

In [11]:
pd.read_sql("""
SELECT name AS planets_without_rings, rings
FROM planets
WHERE rings = 0;
""", conn)
# Notice how code below is more consice and narrow!
# Below is correct, because COUNT = counts the number of records that meet a certain condition

planets_without_rings  rings
0               Mercury      0
1                 Venus      0
2                 Earth      0
3                  Mars      0
4               Jupiter      0

In [12]:
pd.read_sql("""
SELECT COUNT(*) AS planets_without_rings
FROM planets
WHERE rings = 0;
""", conn)

planets_without_rings
0                      5

### 10. Select the name of all planets, along with a value `has_rings` that returns "Yes" if the planet does have rings, and "No" if it does not

>



In [14]:
pd.read_sql("""
SELECT name, rings AS has_rings,
  CASE rings
    WHEN "1" then "Yes"
    When "0" then "NO"
  END AS has_rings
FROM planets
""", conn)
# Forgot comma at end of 2nd line.

name has_rings has_rings
0  Mercury         0        NO
1    Venus         0        NO
2    Earth         0        NO
3     Mars         0        NO
4  Jupiter         0        NO
5   Saturn         1       Yes
6   Uranus         1       Yes
7  Neptune         1       Yes

## Summary

Congratulations! NASA is one step closer to embarking upon its mission to Mars. In this lab, You practiced writing `SELECT` statements that query a single table to get specific information. You also used other clauses and specified column names to cherry-pick the data we wanted to retrieve.